<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="syntenet">
<title>Inference and Analysis of Synteny Networks • syntenet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.2/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Inference and Analysis of Synteny Networks">
<meta property="og:description" content="syntenet">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-#0092ac"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">syntenet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/syntenet.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/almeidasilvaf/syntenet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="syntenet_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><script src="syntenet_files/d3-4.5.0/d3.min.js"></script><script src="syntenet_files/forceNetwork-binding-0.4/forceNetwork.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Inference and Analysis of Synteny Networks</h1>
                        <h4 data-toc-skip class="author">Fabricio
Almeida-Silva</h4>
            <address class="author_afil">
      VIB-UGent Center for Plant Systems Biology, Ghent,
BelgiumDepartment of Plant Biotechnology and Bioinformatics, Ghent
University, Ghent,
Belgium<br><h4 data-toc-skip class="author">Tao
Zhao</h4>
            <address class="author_afil">
      State Key Laboratory of Crop Stress Biology for Arid Areas/Shaanxi
Key Laboratory of Apple, College of Horticulture, Northwest A&amp;F
University, Yangling,
China<br><h4 data-toc-skip class="author">Kristian
K Ullrich</h4>
            <address class="author_afil">
      Department of Evolutionary Biology, Max Planck Institute For
Evolutionary Biology, Ploen,
Germany<br><h4 data-toc-skip class="author">Yves
Van de Peer</h4>
            <address class="author_afil">
      VIB-UGent Center for Plant Systems Biology, Ghent,
BelgiumDepartment of Plant Biotechnology and Bioinformatics, Ghent
University, Ghent, BelgiumCollege of Horticulture, Academy for Advanced
Interdisciplinary Studies, Nanjing Agricultural University, Nanjing,
ChinaCenter for Microbial Ecology and Genomics, Department of
Biochemistry, Genetics and Microbiology, University of Pretoria,
Pretoria, South Africa<br><small class="dont-index">Source: <a href="https://github.com/almeidasilvaf/syntenet/blob/HEAD/vignettes/syntenet.Rmd" class="external-link"><code>vignettes/syntenet.Rmd</code></a></small>
      <div class="d-none name"><code>syntenet.Rmd</code></div>
    </address>
</address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The analysis of synteny (i.e., conserved gene content and order in a
genomic segment across species) can help understand the trajectory of
duplicated genes through evolution. In particular, synteny analyses are
widely used to investigate the evolution of genes derived from
whole-genome duplication (WGD) events, as they can reveal genomic
rearrangements that happened following the duplication of all
chromosomes. However, synteny analysis are typically performed in a
pairwise manner, which can be difficult to explore and interpret when
comparing several species. To understand global patterns of synteny,
<span class="citation">Zhao and Schranz (2017)</span> proposed a
network-based approach to analyze synteny. In synteny networks, genes in
a given syntenic block are represented as nodes connected by an edge.
Synteny networks have been used to explore, among others, global synteny
patterns in mammalian and angiosperm genomes <span class="citation">(Zhao and Schranz 2019)</span>, the evolution of
MADS-box transcription factors <span class="citation">(Zhao et al.
2017)</span>, and infer a microsynteny-based phylogeny for angiosperms
<span class="citation">(Zhao et al. 2021)</span>. <em><a href="https://bioconductor.org/packages/3.15/syntenet" class="external-link">syntenet</a></em>
is a package that can be used to infer synteny networks from protein
sequences and perform downstream network analyses that include:</p>
<ul>
<li><p><strong>Network clustering</strong> using the Infomap
algorithm;</p></li>
<li><p><strong>Phylogenomic profiling</strong>, which consists in
identifying which species contain which clusters. This analysis can
reveal highly conserved synteny clusters and taxon-specific ones (e.g.,
family- and order-specific clusters);</p></li>
<li><p><strong>Microsynteny-based phylogeny reconstruction</strong> with
maximum likelihood, which can be achieved by inferring a phylogeny from
a binary matrix of phylogenomic profiles with IQTREE2.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.15/syntenet" class="external-link">syntenet</a></em>
can be installed from Bioconductor with the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"syntenet"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load package after installation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/almeidasilvaf/syntenet" class="external-link">syntenet</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-description">Data description<a class="anchor" aria-label="anchor" href="#data-description"></a>
</h2>
<p>For this vignette, we will use the proteomes and gene annotation of
the algae species <em>Ostreococcus lucimarinus</em> and <em>Ostreococcus
sp RCC809</em>, which were obtained from Pico-PLAZA 3.0 <span class="citation">(Vandepoele et al. 2013)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Protein sequences</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">proteomes</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">proteomes</span><span class="op">)</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; AAStringSet object of length 1901:</span></span>
<span><span class="co">#&gt;        width seq                                            names               </span></span>
<span><span class="co">#&gt;    [1]   911 MTTMADERASIARVSVVKYGAI...VQLYTYPGSTNDPNFLLKLA* OL01G00010</span></span>
<span><span class="co">#&gt;    [2]   789 MGGRRCFCSRSSPVGVGAPAPA...PPQCGADIEAGSEPPPDKCG* OL01G00020</span></span>
<span><span class="co">#&gt;    [3]   618 MTRAKDAIVVDDGNDDDDDDDD...RDASASLALALAFSSEESVV* OL01G00030</span></span>
<span><span class="co">#&gt;    [4]   547 MPTKAQCWVVSYARVRDGASRS...TGSVSARASIFGEQASFRKA* OL01G00040</span></span>
<span><span class="co">#&gt;    [5]   319 MFTASHTTSKVTLRARVATQPR...HNGMALWRETTPKDSLIPAL* OL01G00050</span></span>
<span><span class="co">#&gt;    ...   ... ...</span></span>
<span><span class="co">#&gt; [1897]   106 MAANDGETKLPEDGWIQPCFAC...RAIVDQVGGEHLKGSLMPIE* OL03G05910</span></span>
<span><span class="co">#&gt; [1898]    70 RMGIVKLATDGSVWVHSPIELD...QQWKDAYPGATLYACPGLKSK OL03G05920</span></span>
<span><span class="co">#&gt; [1899]   680 MDDAHDARWATTSARDGERARA...RSVGPSASDKILEALFPVAD* OL03G05930</span></span>
<span><span class="co">#&gt; [1900]   179 MRAVRERSKANLAARVKEEATR...ELERTRELFARARVRAYECI* OL03G05940</span></span>
<span><span class="co">#&gt; [1901]    83 MFVREARRAIPRFIKDPPQAFH...ESGDVRSVEGEVCGAVLVDE* OL03G05950</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; AAStringSet object of length 1433:</span></span>
<span><span class="co">#&gt;        width seq                                            names               </span></span>
<span><span class="co">#&gt;    [1]   274 MASTTGSAARRVFVDVEKTVNG...DVLSLGQGSLSGESSSSDEE* ORCC809_01G00010</span></span>
<span><span class="co">#&gt;    [2]   175 MDQMRAANAQRSYLLFFVLFFL...SRRLLGRLDSEHTDLHPSWR* ORCC809_01G00020</span></span>
<span><span class="co">#&gt;    [3]   403 MTAPRVRASRRATATAAATVTA...LTERDLRYMEPKATIEEWMG* ORCC809_01G00030</span></span>
<span><span class="co">#&gt;    [4]   217 MTIDADGDDTLAPHAPAHGEVS...LIRLRGVEKTPTVDPPPPPP* ORCC809_01G00040</span></span>
<span><span class="co">#&gt;    [5]  1691 RIEADEKSLLVFGKESPVRTAC...VRMGNNVVTSRYASSESEEDV ORCC809_01G00050</span></span>
<span><span class="co">#&gt;    ...   ... ...</span></span>
<span><span class="co">#&gt; [1429]   428 MVDANATTQTFVLEAEQELRVE...DLPSNVLLVGNLKWLGEDGK* ORCC809_03G02980</span></span>
<span><span class="co">#&gt; [1430]   378 MSVPRTTLRRIPLGNARDVLVT...ETLKAIDAVHAQCRDPCIAT* ORCC809_03G02990</span></span>
<span><span class="co">#&gt; [1431]  1156 MRATSAPSIVSFVARVACLFVA...CAFGTSLASFVVERARRLEN* ORCC809_03G03000</span></span>
<span><span class="co">#&gt; [1432]   541 MAITVFLTDHGRRASALTFLVV...GFGVGAVKFMLAPEMVKSLA* ORCC809_03G03010</span></span>
<span><span class="co">#&gt; [1433]   289 MSLSSLRSFSRSISSAPGGRSC...EPEEPEPEEPEPEEPEEPEP* ORCC809_03G03020</span></span>
<span></span>
<span><span class="co"># Annotation (ranges)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt; GRangesList object of length 2:</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; GRanges object with 1903 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames        ranges strand |     type          ID        Name</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1]    Chr_1      939-3671      - |     gene  OL01G00010  OL01G00010</span></span>
<span><span class="co">#&gt;      [2]    Chr_1     3907-6927      + |     gene  OL01G00020  OL01G00020</span></span>
<span><span class="co">#&gt;      [3]    Chr_1     7085-9160      + |     gene  OL01G00030  OL01G00030</span></span>
<span><span class="co">#&gt;      [4]    Chr_1    9830-11480      + |     gene  OL01G00040  OL01G00040</span></span>
<span><span class="co">#&gt;      [5]    Chr_1   11467-12599      - |     gene  OL01G00050  OL01G00050</span></span>
<span><span class="co">#&gt;      ...      ...           ...    ... .      ...         ...         ...</span></span>
<span><span class="co">#&gt;   [1899]    Chr_3 977435-977752      - |     gene  OL03G05910  OL03G05910</span></span>
<span><span class="co">#&gt;   [1900]    Chr_3 978702-978911      - |     gene  OL03G05920  OL03G05920</span></span>
<span><span class="co">#&gt;   [1901]    Chr_3 979281-981320      - |     gene  OL03G05930  OL03G05930</span></span>
<span><span class="co">#&gt;   [1902]    Chr_3 981778-982314      + |     gene  OL03G05940  OL03G05940</span></span>
<span><span class="co">#&gt;   [1903]    Chr_3 982498-982746      + |     gene  OL03G05950  OL03G05950</span></span>
<span><span class="co">#&gt;              gene_id</span></span>
<span><span class="co">#&gt;          &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1]  OL01G00010</span></span>
<span><span class="co">#&gt;      [2]  OL01G00020</span></span>
<span><span class="co">#&gt;      [3]  OL01G00030</span></span>
<span><span class="co">#&gt;      [4]  OL01G00040</span></span>
<span><span class="co">#&gt;      [5]  OL01G00050</span></span>
<span><span class="co">#&gt;      ...         ...</span></span>
<span><span class="co">#&gt;   [1899]  OL03G05910</span></span>
<span><span class="co">#&gt;   [1900]  OL03G05920</span></span>
<span><span class="co">#&gt;   [1901]  OL03G05930</span></span>
<span><span class="co">#&gt;   [1902]  OL03G05940</span></span>
<span><span class="co">#&gt;   [1903]  OL03G05950</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 6 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; GRanges object with 1433 ranges and 4 metadata columns:</span></span>
<span><span class="co">#&gt;          seqnames        ranges strand |     type               ID</span></span>
<span><span class="co">#&gt;             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;      &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1]    chr_1      321-1142      - |     gene ORCC809_01G00010</span></span>
<span><span class="co">#&gt;      [2]    chr_1     1463-2089      + |     gene ORCC809_01G00020</span></span>
<span><span class="co">#&gt;      [3]    chr_1     2162-3370      - |     gene ORCC809_01G00030</span></span>
<span><span class="co">#&gt;      [4]    chr_1     3774-4424      - |     gene ORCC809_01G00040</span></span>
<span><span class="co">#&gt;      [5]    chr_1     4693-9924      - |     gene ORCC809_01G00050</span></span>
<span><span class="co">#&gt;      ...      ...           ...    ... .      ...              ...</span></span>
<span><span class="co">#&gt;   [1429]    chr_3 504915-506198      - |     gene ORCC809_03G02980</span></span>
<span><span class="co">#&gt;   [1430]    chr_3 506377-507510      + |     gene ORCC809_03G02990</span></span>
<span><span class="co">#&gt;   [1431]    chr_3 507856-511323      + |     gene ORCC809_03G03000</span></span>
<span><span class="co">#&gt;   [1432]    chr_3 511533-513155      - |     gene ORCC809_03G03010</span></span>
<span><span class="co">#&gt;   [1433]    chr_3 513841-514707      + |     gene ORCC809_03G03020</span></span>
<span><span class="co">#&gt;                      Name          gene_id</span></span>
<span><span class="co">#&gt;               &lt;character&gt;      &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1] ORCC809_01G00010 ORCC809_01G00010</span></span>
<span><span class="co">#&gt;      [2] ORCC809_01G00020 ORCC809_01G00020</span></span>
<span><span class="co">#&gt;      [3] ORCC809_01G00030 ORCC809_01G00030</span></span>
<span><span class="co">#&gt;      [4] ORCC809_01G00040 ORCC809_01G00040</span></span>
<span><span class="co">#&gt;      [5] ORCC809_01G00050 ORCC809_01G00050</span></span>
<span><span class="co">#&gt;      ...              ...              ...</span></span>
<span><span class="co">#&gt;   [1429] ORCC809_03G02980 ORCC809_03G02980</span></span>
<span><span class="co">#&gt;   [1430] ORCC809_03G02990 ORCC809_03G02990</span></span>
<span><span class="co">#&gt;   [1431] ORCC809_03G03000 ORCC809_03G03000</span></span>
<span><span class="co">#&gt;   [1432] ORCC809_03G03010 ORCC809_03G03010</span></span>
<span><span class="co">#&gt;   [1433] ORCC809_03G03020 ORCC809_03G03020</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 6 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="importing-data-to-the-r-session">Importing data to the R session<a class="anchor" aria-label="anchor" href="#importing-data-to-the-r-session"></a>
</h2>
<p>To detect synteny and infer synteny networks, <em><a href="https://bioconductor.org/packages/3.15/syntenet" class="external-link">syntenet</a></em>
requires two objects as input:</p>
<ul>
<li>
<strong>seq:</strong> A list of <code>AAStringSet</code> objects
containing the translated sequences of primary transcripts for each
species.</li>
<li>
<strong>annotation:</strong> A <code>GRangesList</code> or
<code>CompressedGRangesList</code> object containing the coordinates for
the genes in <strong>seq</strong>.</li>
</ul>
<p>If you have whole-genome protein sequences in FASTA files, store all
FASTA files in the same directory and use the function
<code><a href="../reference/fasta2AAStringSetlist.html">fasta2AAStringSetlist()</a></code> to read all FASTA files into a list
of <code>AAStringSet</code> objects.</p>
<p>Likewise, if you have gene annotation in GFF/GFF3/GTF files, store
all files in the same directory and use the function
<code><a href="../reference/gff2GRangesList.html">gff2GRangesList()</a></code> to read all GFF/GFF3/GTF files into a
<code>GRangesList object</code>.</p>
<p>For a demonstration, we will read example FASTA and GFF3 files stored
in subdirectories named <strong>sequences/</strong> and
<strong>annotation/</strong>, which are located in the
<code>extdata/</code> directory of this package.</p>
<div class="section level3">
<h3 id="from-fasta-files-to-a-list-of-aastringset-objects">From FASTA files to a list of <code>AAStringSet</code> objects<a class="anchor" aria-label="anchor" href="#from-fasta-files-to-a-list-of-aastringset-objects"></a>
</h3>
<p>Here is how you can use <code><a href="../reference/fasta2AAStringSetlist.html">fasta2AAStringSetlist()</a></code> to read
FASTA files in a directory as a list of <code>AAStringSet</code>
objects.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to directory containing FASTA files</span></span>
<span><span class="va">fasta_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sequences"</span>, package <span class="op">=</span> <span class="st">"syntenet"</span><span class="op">)</span></span>
<span><span class="va">fasta_dir</span></span>
<span><span class="co">#&gt; [1] "/__w/_temp/Library/syntenet/extdata/sequences"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">fasta_dir</span><span class="op">)</span> <span class="co"># see the contents of the directory</span></span>
<span><span class="co">#&gt; [1] "Olucimarinus.fa.gz" "OspRCC809.fa.gz"</span></span>
<span></span>
<span><span class="co"># Read all FASTA files in `fasta_dir`</span></span>
<span><span class="va">aastringsetlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fasta2AAStringSetlist.html">fasta2AAStringSetlist</a></span><span class="op">(</span><span class="va">fasta_dir</span><span class="op">)</span></span>
<span><span class="va">aastringsetlist</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; AAStringSet object of length 100:</span></span>
<span><span class="co">#&gt;       width seq                                             names               </span></span>
<span><span class="co">#&gt;   [1]   911 MTTMADERASIARVSVVKYGAI...DVQLYTYPGSTNDPNFLLKLA* OL01G00010</span></span>
<span><span class="co">#&gt;   [2]   789 MGGRRCFCSRSSPVGVGAPAPA...FPPQCGADIEAGSEPPPDKCG* OL01G00020</span></span>
<span><span class="co">#&gt;   [3]   618 MTRAKDAIVVDDGNDDDDDDDD...DRDASASLALALAFSSEESVV* OL01G00030</span></span>
<span><span class="co">#&gt;   [4]   547 MPTKAQCWVVSYARVRDGASRS...VTGSVSARASIFGEQASFRKA* OL01G00040</span></span>
<span><span class="co">#&gt;   [5]   319 MFTASHTTSKVTLRARVATQPR...LHNGMALWRETTPKDSLIPAL* OL01G00050</span></span>
<span><span class="co">#&gt;   ...   ... ...</span></span>
<span><span class="co">#&gt;  [96]   476 MVPARNFLDGANAREVELDRVV...VMRKLREPDSVARLAGQTGVR* OL01G00960</span></span>
<span><span class="co">#&gt;  [97]   771 MARHRGTRGGWNATTTEGGDGR...SIPDDGFDESSSVSASTIDGF* OL01G00970</span></span>
<span><span class="co">#&gt;  [98]   494 MDSEFWGCVIPAGRAVRVEVAT...FIKSRKDLFTIDGAYVRLVKK* OL01G00980</span></span>
<span><span class="co">#&gt;  [99]   264 VRAIVGATTRIQTRAPPRANHR...DWSFISDEFQDDASDSEVIDR* OL01G00990</span></span>
<span><span class="co">#&gt; [100]   565 MQLDAFRKATVKGVATRVGGAD...QLADLLRKNMGVPAKFIDAQN* OL01G01000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; AAStringSet object of length 100:</span></span>
<span><span class="co">#&gt;       width seq                                             names               </span></span>
<span><span class="co">#&gt;   [1]   274 MASTTGSAARRVFVDVEKTVNG...WDVLSLGQGSLSGESSSSDEE* ORCC809_01G00010</span></span>
<span><span class="co">#&gt;   [2]   175 MDQMRAANAQRSYLLFFVLFFL...SSRRLLGRLDSEHTDLHPSWR* ORCC809_01G00020</span></span>
<span><span class="co">#&gt;   [3]   403 MTAPRVRASRRATATAAATVTA...ALTERDLRYMEPKATIEEWMG* ORCC809_01G00030</span></span>
<span><span class="co">#&gt;   [4]   217 MTIDADGDDTLAPHAPAHGEVS...SLIRLRGVEKTPTVDPPPPPP* ORCC809_01G00040</span></span>
<span><span class="co">#&gt;   [5]  1691 RIEADEKSLLVFGKESPVRTAC...SVRMGNNVVTSRYASSESEEDV ORCC809_01G00050</span></span>
<span><span class="co">#&gt;   ...   ... ...</span></span>
<span><span class="co">#&gt;  [96]   357 MSRGLADNWDDAEGYYCARIGE...TVNEALQHPFIVERIRTTAPN* ORCC809_01G00960</span></span>
<span><span class="co">#&gt;  [97]   164 MAMDSFRSAPRSRRRVEATSRE...SKPVKPVREPVRMVEASTGAH* ORCC809_01G00970</span></span>
<span><span class="co">#&gt;  [98]    85 MPEGTVFIGNIPYDATESSLTE...NLNAREYNGRQLRVDHAETMKG ORCC809_01G00980</span></span>
<span><span class="co">#&gt;  [99]   229 MKGGGGASGAAASANGNGAVGG...PDQRAQVEYLRQLAAQQGMVR* ORCC809_01G00990</span></span>
<span><span class="co">#&gt; [100]   103 RKAGGERWEDSSLAEWPENDFR...EMAGKFIGNRPVKLRKSAWNER ORCC809_01G01000</span></span></code></pre></div>
<p>And that’s it! Now you have a list of <code>AAStringSet</code>
objects.</p>
</div>
<div class="section level3">
<h3 id="from-gffgtf-files-to-a-grangeslist-object">From GFF/GTF files to a <code>GRangesList</code> object<a class="anchor" aria-label="anchor" href="#from-gffgtf-files-to-a-grangeslist-object"></a>
</h3>
<p>Here is how you can use <code><a href="../reference/gff2GRangesList.html">gff2GRangesList()</a></code> to read
GFF/GFF3/GTF files in a directory as a <code>GRangesList</code>
object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Path to directory containing FASTA files</span></span>
<span><span class="va">gff_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"annotation"</span>, package <span class="op">=</span> <span class="st">"syntenet"</span><span class="op">)</span></span>
<span><span class="va">gff_dir</span></span>
<span><span class="co">#&gt; [1] "/__w/_temp/Library/syntenet/extdata/annotation"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">gff_dir</span><span class="op">)</span> <span class="co"># see the contents of the directory</span></span>
<span><span class="co">#&gt; [1] "Olucimarinus.gff3.gz" "OspRCC809.gff3.gz"</span></span>
<span></span>
<span><span class="co"># Read all FASTA files in `fasta_dir`</span></span>
<span><span class="va">grangeslist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gff2GRangesList.html">gff2GRangesList</a></span><span class="op">(</span><span class="va">gff_dir</span><span class="op">)</span></span>
<span><span class="va">grangeslist</span></span>
<span><span class="co">#&gt; GRangesList object of length 2:</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; GRanges object with 100 ranges and 7 metadata columns:</span></span>
<span><span class="co">#&gt;         seqnames        ranges strand |      source     type     score</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    Chr_1      939-3671      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [2]    Chr_1     3907-6927      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [3]    Chr_1     7085-9160      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [4]    Chr_1    9830-11480      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [5]    Chr_1   11467-12599      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     ...      ...           ...    ... .         ...      ...       ...</span></span>
<span><span class="co">#&gt;    [96]    Chr_1 170975-172402      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [97]    Chr_1 172445-174757      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [98]    Chr_1 175358-176839      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [99]    Chr_1 176901-177692      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;   [100]    Chr_1 177742-179436      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;             phase          ID        Name     gene_id</span></span>
<span><span class="co">#&gt;         &lt;integer&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;     [1]      &lt;NA&gt;  OL01G00010  OL01G00010  OL01G00010</span></span>
<span><span class="co">#&gt;     [2]      &lt;NA&gt;  OL01G00020  OL01G00020  OL01G00020</span></span>
<span><span class="co">#&gt;     [3]      &lt;NA&gt;  OL01G00030  OL01G00030  OL01G00030</span></span>
<span><span class="co">#&gt;     [4]      &lt;NA&gt;  OL01G00040  OL01G00040  OL01G00040</span></span>
<span><span class="co">#&gt;     [5]      &lt;NA&gt;  OL01G00050  OL01G00050  OL01G00050</span></span>
<span><span class="co">#&gt;     ...       ...         ...         ...         ...</span></span>
<span><span class="co">#&gt;    [96]      &lt;NA&gt;  OL01G00960  OL01G00960  OL01G00960</span></span>
<span><span class="co">#&gt;    [97]      &lt;NA&gt;  OL01G00970  OL01G00970  OL01G00970</span></span>
<span><span class="co">#&gt;    [98]      &lt;NA&gt;  OL01G00980  OL01G00980  OL01G00980</span></span>
<span><span class="co">#&gt;    [99]      &lt;NA&gt;  OL01G00990  OL01G00990  OL01G00990</span></span>
<span><span class="co">#&gt;   [100]      &lt;NA&gt;  OL01G01000  OL01G01000  OL01G01000</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; GRanges object with 100 ranges and 7 metadata columns:</span></span>
<span><span class="co">#&gt;         seqnames        ranges strand |      source     type     score</span></span>
<span><span class="co">#&gt;            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt;     [1]    chr_1      321-1142      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [2]    chr_1     1463-2089      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [3]    chr_1     2162-3370      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [4]    chr_1     3774-4424      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     [5]    chr_1     4693-9924      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;     ...      ...           ...    ... .         ...      ...       ...</span></span>
<span><span class="co">#&gt;    [96]    chr_1 165459-166529      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [97]    chr_1 166654-167213      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [98]    chr_1 167296-167550      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;    [99]    chr_1 167542-168228      + | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;   [100]    chr_1 168639-168947      - | rtracklayer     gene        NA</span></span>
<span><span class="co">#&gt;             phase               ID             Name          gene_id</span></span>
<span><span class="co">#&gt;         &lt;integer&gt;      &lt;character&gt;      &lt;character&gt;      &lt;character&gt;</span></span>
<span><span class="co">#&gt;     [1]      &lt;NA&gt; ORCC809_01G00010 ORCC809_01G00010 ORCC809_01G00010</span></span>
<span><span class="co">#&gt;     [2]      &lt;NA&gt; ORCC809_01G00020 ORCC809_01G00020 ORCC809_01G00020</span></span>
<span><span class="co">#&gt;     [3]      &lt;NA&gt; ORCC809_01G00030 ORCC809_01G00030 ORCC809_01G00030</span></span>
<span><span class="co">#&gt;     [4]      &lt;NA&gt; ORCC809_01G00040 ORCC809_01G00040 ORCC809_01G00040</span></span>
<span><span class="co">#&gt;     [5]      &lt;NA&gt; ORCC809_01G00050 ORCC809_01G00050 ORCC809_01G00050</span></span>
<span><span class="co">#&gt;     ...       ...              ...              ...              ...</span></span>
<span><span class="co">#&gt;    [96]      &lt;NA&gt; ORCC809_01G00960 ORCC809_01G00960 ORCC809_01G00960</span></span>
<span><span class="co">#&gt;    [97]      &lt;NA&gt; ORCC809_01G00970 ORCC809_01G00970 ORCC809_01G00970</span></span>
<span><span class="co">#&gt;    [98]      &lt;NA&gt; ORCC809_01G00980 ORCC809_01G00980 ORCC809_01G00980</span></span>
<span><span class="co">#&gt;    [99]      &lt;NA&gt; ORCC809_01G00990 ORCC809_01G00990 ORCC809_01G00990</span></span>
<span><span class="co">#&gt;   [100]      &lt;NA&gt; ORCC809_01G01000 ORCC809_01G01000 ORCC809_01G01000</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
<p>And now you have a <code>GRangesList</code> object.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a>
</h2>
<p>The first part of the pipeline consists in processing the data to
make it match a standard structure. However, before processing the data
for synteny detection, you must use the function
<code><a href="../reference/check_input.html">check_input()</a></code> to check if your data can enter the pipeline.
This function checks the input data for 3 required conditions:</p>
<ol style="list-style-type: decimal">
<li><p>Names of <strong>seq</strong> list (i.e.,
<code>names(seq)</code>) match the names of <strong>annotation</strong>
<code>GRangesList</code>/<code>CompressedGRangesList</code> (i.e.,
<code>names(annotation)</code>)</p></li>
<li><p>For each species (list elements), the number of sequences in
<strong>seq</strong> is not greater than the number of genes in
<strong>annotation</strong>. This is a way to ensure users do not input
the translated sequences for multiple isoforms of the same gene
(generated by alternative splicing). Ideally, the number of sequences in
<strong>seq</strong> should be equal to the number of genes in
<strong>annotation</strong>, but this may not always stand true because
of non-protein-coding genes.</p></li>
<li><p>For each species, sequence names (i.e.,
<code>names(seq[[x]])</code>, equivalent to FASTA headers) match gene
names in <code>annotation</code>.</p></li>
</ol>
<p>Let’s check if the example data sets satisfy these 3 criteria:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_input.html">check_input</a></span><span class="op">(</span><span class="va">proteomes</span>, <span class="va">annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>As you can see, the data passed the checks. Now, let’s process them
with the function <code><a href="../reference/process_input.html">process_input()</a></code>.</p>
<p>This function processes the input sequences and annotation to:</p>
<ol style="list-style-type: decimal">
<li><p>Remove whitespace and anything after it in sequence names (i.e.,
<code>names(seq[[x]])</code>, which is equivalent to FASTA headers), if
there is any.</p></li>
<li><p>Remove period followed by number at the end of sequence names,
which typically indicates different isoforms of the same gene (e.g.,
Arabidopsis thaliana’s transcript AT1G01010.1, which belongs to gene
AT1G01010).</p></li>
<li><p>Add a unique species identifier to sequence names. The species
identifier consists of the first 3-5 strings of the element name. For
instance, if the first element of the <strong>seq</strong> list is named
“Athaliana”, each sequence in it will have an identifier “Atha_” added
to the beginning of each gene name (e.g., Atha_AT1G01010).</p></li>
<li><p>If sequences have an asterisk (*) representing stop codon, remove
it.</p></li>
<li><p>Add a unique species identifier (same as above) to gene and
chromosome names of each element of the <strong>annotation</strong>
<code>GRangesList</code>/<code>CompressedGRangesList</code>.</p></li>
<li><p>Filter each element of the <strong>annotation</strong>
<code>GRangesList</code>/<code>CompressedGRangesList</code> to keep only
seqnames, ranges, and gene ID.</p></li>
</ol>
<p>Let’s process our input data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_input.html">process_input</a></span><span class="op">(</span><span class="va">proteomes</span>, <span class="va">annotation</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Looking at the processed data</span></span>
<span><span class="va">pdata</span><span class="op">$</span><span class="va">seq</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; AAStringSet object of length 1901:</span></span>
<span><span class="co">#&gt;        width seq                                            names               </span></span>
<span><span class="co">#&gt;    [1]   910 MTTMADERASIARVSVVKYGAI...DVQLYTYPGSTNDPNFLLKLA Olu_OL01G00010</span></span>
<span><span class="co">#&gt;    [2]   788 MGGRRCFCSRSSPVGVGAPAPA...FPPQCGADIEAGSEPPPDKCG Olu_OL01G00020</span></span>
<span><span class="co">#&gt;    [3]   617 MTRAKDAIVVDDGNDDDDDDDD...DRDASASLALALAFSSEESVV Olu_OL01G00030</span></span>
<span><span class="co">#&gt;    [4]   546 MPTKAQCWVVSYARVRDGASRS...VTGSVSARASIFGEQASFRKA Olu_OL01G00040</span></span>
<span><span class="co">#&gt;    [5]   318 MFTASHTTSKVTLRARVATQPR...LHNGMALWRETTPKDSLIPAL Olu_OL01G00050</span></span>
<span><span class="co">#&gt;    ...   ... ...</span></span>
<span><span class="co">#&gt; [1897]   105 MAANDGETKLPEDGWIQPCFAC...LRAIVDQVGGEHLKGSLMPIE Olu_OL03G05910</span></span>
<span><span class="co">#&gt; [1898]    69 RMGIVKLATDGSVWVHSPIELD...AQQWKDAYPGATLYACPGLKS Olu_OL03G05920</span></span>
<span><span class="co">#&gt; [1899]   679 MDDAHDARWATTSARDGERARA...ARSVGPSASDKILEALFPVAD Olu_OL03G05930</span></span>
<span><span class="co">#&gt; [1900]   178 MRAVRERSKANLAARVKEEATR...LELERTRELFARARVRAYECI Olu_OL03G05940</span></span>
<span><span class="co">#&gt; [1901]    82 MFVREARRAIPRFIKDPPQAFH...QESGDVRSVEGEVCGAVLVDE Olu_OL03G05950</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; AAStringSet object of length 1433:</span></span>
<span><span class="co">#&gt;        width seq                                            names               </span></span>
<span><span class="co">#&gt;    [1]   273 MASTTGSAARRVFVDVEKTVNG...WDVLSLGQGSLSGESSSSDEE Osp_ORCC809_01G00010</span></span>
<span><span class="co">#&gt;    [2]   174 MDQMRAANAQRSYLLFFVLFFL...SSRRLLGRLDSEHTDLHPSWR Osp_ORCC809_01G00020</span></span>
<span><span class="co">#&gt;    [3]   402 MTAPRVRASRRATATAAATVTA...ALTERDLRYMEPKATIEEWMG Osp_ORCC809_01G00030</span></span>
<span><span class="co">#&gt;    [4]   216 MTIDADGDDTLAPHAPAHGEVS...SLIRLRGVEKTPTVDPPPPPP Osp_ORCC809_01G00040</span></span>
<span><span class="co">#&gt;    [5]  1690 RIEADEKSLLVFGKESPVRTAC...SVRMGNNVVTSRYASSESEED Osp_ORCC809_01G00050</span></span>
<span><span class="co">#&gt;    ...   ... ...</span></span>
<span><span class="co">#&gt; [1429]   427 MVDANATTQTFVLEAEQELRVE...GDLPSNVLLVGNLKWLGEDGK Osp_ORCC809_03G02980</span></span>
<span><span class="co">#&gt; [1430]   377 MSVPRTTLRRIPLGNARDVLVT...KETLKAIDAVHAQCRDPCIAT Osp_ORCC809_03G02990</span></span>
<span><span class="co">#&gt; [1431]  1155 MRATSAPSIVSFVARVACLFVA...ACAFGTSLASFVVERARRLEN Osp_ORCC809_03G03000</span></span>
<span><span class="co">#&gt; [1432]   540 MAITVFLTDHGRRASALTFLVV...PGFGVGAVKFMLAPEMVKSLA Osp_ORCC809_03G03010</span></span>
<span><span class="co">#&gt; [1433]   288 MSLSSLRSFSRSISSAPGGRSC...EEPEEPEPEEPEPEEPEEPEP Osp_ORCC809_03G03020</span></span>
<span><span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span></span>
<span><span class="co">#&gt; $Olucimarinus</span></span>
<span><span class="co">#&gt; GRanges object with 1903 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;           seqnames        ranges strand |           gene</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |    &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1] Olu_Chr_1      939-3671      * | Olu_OL01G00010</span></span>
<span><span class="co">#&gt;      [2] Olu_Chr_1     3907-6927      * | Olu_OL01G00020</span></span>
<span><span class="co">#&gt;      [3] Olu_Chr_1     7085-9160      * | Olu_OL01G00030</span></span>
<span><span class="co">#&gt;      [4] Olu_Chr_1    9830-11480      * | Olu_OL01G00040</span></span>
<span><span class="co">#&gt;      [5] Olu_Chr_1   11467-12599      * | Olu_OL01G00050</span></span>
<span><span class="co">#&gt;      ...       ...           ...    ... .            ...</span></span>
<span><span class="co">#&gt;   [1899] Olu_Chr_3 977435-977752      * | Olu_OL03G05910</span></span>
<span><span class="co">#&gt;   [1900] Olu_Chr_3 978702-978911      * | Olu_OL03G05920</span></span>
<span><span class="co">#&gt;   [1901] Olu_Chr_3 979281-981320      * | Olu_OL03G05930</span></span>
<span><span class="co">#&gt;   [1902] Olu_Chr_3 981778-982314      * | Olu_OL03G05940</span></span>
<span><span class="co">#&gt;   [1903] Olu_Chr_3 982498-982746      * | Olu_OL03G05950</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $OspRCC809</span></span>
<span><span class="co">#&gt; GRanges object with 1433 ranges and 1 metadata column:</span></span>
<span><span class="co">#&gt;           seqnames        ranges strand |                 gene</span></span>
<span><span class="co">#&gt;              &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |          &lt;character&gt;</span></span>
<span><span class="co">#&gt;      [1] Osp_chr_1      321-1142      * | Osp_ORCC809_01G00010</span></span>
<span><span class="co">#&gt;      [2] Osp_chr_1     1463-2089      * | Osp_ORCC809_01G00020</span></span>
<span><span class="co">#&gt;      [3] Osp_chr_1     2162-3370      * | Osp_ORCC809_01G00030</span></span>
<span><span class="co">#&gt;      [4] Osp_chr_1     3774-4424      * | Osp_ORCC809_01G00040</span></span>
<span><span class="co">#&gt;      [5] Osp_chr_1     4693-9924      * | Osp_ORCC809_01G00050</span></span>
<span><span class="co">#&gt;      ...       ...           ...    ... .                  ...</span></span>
<span><span class="co">#&gt;   [1429] Osp_chr_3 504915-506198      * | Osp_ORCC809_03G02980</span></span>
<span><span class="co">#&gt;   [1430] Osp_chr_3 506377-507510      * | Osp_ORCC809_03G02990</span></span>
<span><span class="co">#&gt;   [1431] Osp_chr_3 507856-511323      * | Osp_ORCC809_03G03000</span></span>
<span><span class="co">#&gt;   [1432] Osp_chr_3 511533-513155      * | Osp_ORCC809_03G03010</span></span>
<span><span class="co">#&gt;   [1433] Osp_chr_3 513841-514707      * | Osp_ORCC809_03G03020</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="synteny-network-inference">Synteny network inference<a class="anchor" aria-label="anchor" href="#synteny-network-inference"></a>
</h2>
<p>Now that we have our processed data, we can infer the synteny
network. To detect synteny, we need the tabular output from BLASTp <span class="citation">(Altschul et al. 1997)</span> or similar programs. To
get that, you can use the function <code><a href="../reference/run_diamond.html">run_diamond()</a></code>, which
runs DIAMOND <span class="citation">(Buchfink, Reuter, and Drost
2021)</span> from the R session and automatically parses its output to a
list of data frames <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;strong&gt;Alternative:&lt;/strong&gt; if you want to use a
different program for similarity searches, you can run it on the command
line, save the output in tabular format, and read the tabular output as
a data frame (e.g., with the base R function &lt;code&gt;read.table&lt;/code&gt;).&lt;/p&gt;"><sup>1</sup></a>.</p>
<p>Let’s demonstrate how <code><a href="../reference/run_diamond.html">run_diamond()</a></code> works. Needless to
say, you need to have DIAMOND installed in your machine and in your PATH
to run this function. To check if you have DIAMOND installed, use the
function <code><a href="../reference/diamond_is_installed.html">diamond_is_installed()</a></code> <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Note:&lt;/strong&gt; in the code chunk below, the if
statement is not required. We just added it to make sure that the
function &lt;code&gt;&lt;a href="../reference/run_diamond.html"&gt;run_diamond()&lt;/a&gt;&lt;/code&gt; is only executed if DIAMOND is
installed, to avoid problems when building this vignette in machines
that do not have DIAMOND installed. If you want to reproduce the code in
this vignette and do not have DIAMOND installed, you can use the example
output of &lt;code&gt;&lt;a href="../reference/run_diamond.html"&gt;run_diamond()&lt;/a&gt;&lt;/code&gt; stored in the &lt;em&gt;blast_list&lt;/em&gt;
object (loaded with &lt;code&gt;data(blast_list)&lt;/code&gt;).&lt;/p&gt;'><sup>2</sup></a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">blast_list</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/diamond_is_installed.html">diamond_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">blast_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_diamond.html">run_diamond</a></span><span class="op">(</span>seq <span class="op">=</span> <span class="va">pdata</span><span class="op">$</span><span class="va">seq</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The output of <code><a href="../reference/run_diamond.html">run_diamond()</a></code> is a list of data frames
containing the tabular output of all-vs-all DIAMOND searches. Let’s take
a look at it.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># List names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">blast_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Olucimarinus_Olucimarinus" "Olucimarinus_OspRCC809"   </span></span>
<span><span class="co">#&gt; [3] "OspRCC809_Olucimarinus"    "OspRCC809_OspRCC809"</span></span>
<span></span>
<span><span class="co"># Inspect first data frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">blast_list</span><span class="op">$</span><span class="va">Olucimarinus_Olucimarinus</span><span class="op">)</span></span>
<span><span class="co">#&gt;            query             db perc_identity length mismatches gap_open qstart</span></span>
<span><span class="co">#&gt; 1 Olu_OL01G00010 Olu_OL01G00010         100.0    910          0        0      1</span></span>
<span><span class="co">#&gt; 2 Olu_OL01G00010 Olu_OL01G00020          30.0    494        310       10    427</span></span>
<span><span class="co">#&gt; 3 Olu_OL01G00020 Olu_OL01G00020         100.0    788          0        0      1</span></span>
<span><span class="co">#&gt; 4 Olu_OL01G00020 Olu_OL01G00010          28.7    614        370       17    155</span></span>
<span><span class="co">#&gt; 5 Olu_OL01G00030 Olu_OL01G00030         100.0    617          0        0      1</span></span>
<span><span class="co">#&gt; 6 Olu_OL01G00040 Olu_OL01G00040         100.0    546          0        0      1</span></span>
<span><span class="co">#&gt;   qend tstart tend   evalue bitscore</span></span>
<span><span class="co">#&gt; 1  910      1  910 0.00e+00     1623</span></span>
<span><span class="co">#&gt; 2  894    270  753 1.67e-54      199</span></span>
<span><span class="co">#&gt; 3  788      1  788 0.00e+00     1363</span></span>
<span><span class="co">#&gt; 4  753    334  894 3.98e-56      204</span></span>
<span><span class="co">#&gt; 5  617      1  617 0.00e+00     1046</span></span>
<span><span class="co">#&gt; 6  546      1  546 0.00e+00     1035</span></span></code></pre></div>
<p>Now, we can use this list of DIAMOND data frames to detect synteny.
Here, we reimplemented the popular MCScanX algorithm <span class="citation">(Y. Wang et al. 2012)</span>, originally written in
C++, using the <em><a href="https://CRAN.R-project.org/package=Rcpp" class="external-link">Rcpp</a></em> <span class="citation">(Eddelbuettel and François 2011)</span> framework for R
and C++ integration. This means that <em><a href="https://bioconductor.org/packages/3.15/syntenet" class="external-link">syntenet</a></em>
comes with a native version of the MCScanX algorithm, so you can run
MCScanX in R without having to install it yourself. Amazing, huh?</p>
<p>To detect synteny and infer the synteny network, use the function
<code><a href="../reference/infer_syntenet.html">infer_syntenet()</a></code>. The output is a network represented as a
so-called <strong>edge list</strong> (i.e., a 2-column data frame with
node 1 and node 2 in columns 1 and 2, respectively).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Infer synteny network</span></span>
<span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_syntenet.html">infer_syntenet</a></span><span class="op">(</span><span class="va">blast_list</span>, <span class="va">pdata</span><span class="op">$</span><span class="va">annotation</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">net</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Anchor1              Anchor2</span></span>
<span><span class="co">#&gt; 1 Olu_OL01G00070 Osp_ORCC809_01G06520</span></span>
<span><span class="co">#&gt; 2 Olu_OL01G00100 Osp_ORCC809_01G06480</span></span>
<span><span class="co">#&gt; 3 Olu_OL01G00130 Osp_ORCC809_01G06440</span></span>
<span><span class="co">#&gt; 4 Olu_OL01G00150 Osp_ORCC809_01G06420</span></span>
<span><span class="co">#&gt; 5 Olu_OL01G00160 Osp_ORCC809_01G06410</span></span>
<span><span class="co">#&gt; 6 Olu_OL01G00170 Osp_ORCC809_01G06400</span></span></code></pre></div>
<p>In a synteny network, each row of the edge list represents an anchor
pair. In the edge list above, for example, the genes Olu_OL01G00070 and
Osp_ORCC809_01G06520 are in the same syntenic block.</p>
</div>
<div class="section level2">
<h2 id="phylogenomic-profiling">Phylogenomic profiling<a class="anchor" aria-label="anchor" href="#phylogenomic-profiling"></a>
</h2>
<p>After inferring the synteny network, the first thing you would want
to do is cluster your network and identify which phylogenetic groups are
contained in each cluster. This is what we call <strong>phylogenomic
profiling</strong>. This way, you can identify clade-specific clusters,
and highly conserved clusters, for instance. Here, we will use an
example network of BUSCO genes for 25 eudicot species, which was
obtained from <span class="citation">Zhao and Schranz (2019)</span>.</p>
<p>To obtain the phylogenomic profiles, you first need to cluster your
network. This can be done with <code><a href="../reference/cluster_network.html">cluster_network()</a></code>. <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Friendly tip:&lt;/strong&gt; &lt;em&gt;&lt;a href="https://bioconductor.org/packages/3.15/syntenet" class="external-link"&gt;syntenet&lt;/a&gt;&lt;/em&gt;
uses the &lt;em&gt;Infomap&lt;/em&gt; algorithm to cluster networks, which has been
shown to have the best performance &lt;span class="citation"&gt;(Zhao and
Schranz 2019)&lt;/span&gt;. However, you can use any other network clustering
method implemented in the &lt;em&gt;cluster_&lt;/em&gt; family of functions from the
&lt;em&gt;&lt;a href="https://CRAN.R-project.org/package=igraph" class="external-link"&gt;igraph&lt;/a&gt;&lt;/em&gt;
package by passing the function directly to the &lt;em&gt;clust_function&lt;/em&gt;
parameter (see &lt;code&gt;&lt;a href="../reference/cluster_network.html"&gt;?cluster_network&lt;/a&gt;&lt;/code&gt; for details). Importantly,
the Infomap algorithm (default clustering method) assigns each gene to a
single cluster. However, for some cases (e.g., detection of tandem
arrays), you may want to use an algorithm that allows community overlap
(i.e., a gene can be part of more than one cluster). If this is your
case, we recommend the &lt;em&gt;clique percolation&lt;/em&gt; algorithm, which is
implemented in the R package &lt;em&gt;&lt;a href="https://CRAN.R-project.org/package=CliquePercolation" class="external-link"&gt;CliquePercolation&lt;/a&gt;&lt;/em&gt;
&lt;span class="citation"&gt;(Lange 2021)&lt;/span&gt;.&lt;/p&gt;'><sup>3</sup></a></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example data and take a look at it</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span></span>
<span><span class="co">#&gt;       node1           node2</span></span>
<span><span class="co">#&gt; 1 cca_23646  Lang_109327134</span></span>
<span><span class="co">#&gt; 2 cca_23646  Lang_109328075</span></span>
<span><span class="co">#&gt; 3 cca_23646   Mnot_21394516</span></span>
<span><span class="co">#&gt; 4 cca_23646  Zjuj_107413994</span></span>
<span><span class="co">#&gt; 5 cca_23646 adu_Aradu.8SN53</span></span>
<span><span class="co">#&gt; 6 cca_23646     car_14082.1</span></span>
<span></span>
<span><span class="co"># Cluster network</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_network.html">cluster_network</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Gene Cluster</span></span>
<span><span class="co">#&gt; 1 cca_23646       1</span></span>
<span><span class="co">#&gt; 2 cca_23668       2</span></span>
<span><span class="co">#&gt; 3 cca_32926       3</span></span>
<span><span class="co">#&gt; 4 cca_26186       4</span></span>
<span><span class="co">#&gt; 5 cca_24381       5</span></span>
<span><span class="co">#&gt; 6 cca_24396       6</span></span></code></pre></div>
<p>Now that each gene has been assigned to a cluster, we can identify
the phylogenomic profiles of each cluster. This function returns a
matrix of phylogenomic profiles, with clusters in rows and species in
columns.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Phylogenomic profiling</span></span>
<span><span class="va">profiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylogenomic_profile.html">phylogenomic_profile</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Exploring the output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">profiles</span><span class="op">)</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     adu car cca fve gma hlu jcu Lang lja lus mdo mes Mnot mtr pbr pmu ppe ptr</span></span>
<span><span class="co">#&gt;   1   1   1   1   1   2   0   0    2   1   2   2   1    1   1   3   1   1   1</span></span>
<span><span class="co">#&gt;   2   1   1   1   1   1   0   0    1   0   1   1   1    1   1   1   1   1   1</span></span>
<span><span class="co">#&gt;   3   0   1   1   1   1   0   0    1   1   2   1   1    1   1   1   1   1   1</span></span>
<span><span class="co">#&gt;   4   0   1   1   1   2   0   1    2   1   2   0   2    1   1   1   1   1   0</span></span>
<span><span class="co">#&gt;   5   1   1   1   1   2   0   1    0   0   0   1   1    1   1   1   1   1   1</span></span>
<span><span class="co">#&gt;   6   2   1   2   1   3   0   1    2   0   2   2   1    1   2   2   1   1   2</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;     pvu rco roc tpr van vra Zjuj</span></span>
<span><span class="co">#&gt;   1   1   1   1   1   1   1    1</span></span>
<span><span class="co">#&gt;   2   1   0   0   1   1   1    1</span></span>
<span><span class="co">#&gt;   3   1   1   1   1   0   0    1</span></span>
<span><span class="co">#&gt;   4   1   1   0   1   1   0    1</span></span>
<span><span class="co">#&gt;   5   1   1   1   1   1   2    1</span></span>
<span><span class="co">#&gt;   6   1   1   1   1   0   1    1</span></span></code></pre></div>
<p>As a plot is worth a thousand words (or numbers), you can use the
function <code><a href="../reference/plot_profiles.html">plot_profiles()</a></code> to visualize the phylogenomic
profiles as a heatmap with species in rows and synteny network clusters
in columns. The heatmap generated by this function is highly
customizable by users. Some important remarks are:</p>
<ol style="list-style-type: decimal">
<li><p>You can add a legend for species metadata (e.g., taxonomic
information) by passing a 2-column data frame to the parameter
<em>species_annotation</em>.</p></li>
<li><p>Columns (network clusters) are grouped with Ward’s clustering on
a matrix of distances. The method to compute the distance matrix can be
defined by users in parameters <em>dist_function</em> and
<em>dist_params</em>. By default, it uses the function
<code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">stats::dist()</a></code> with parameter
<code>method = "euclidean"</code>. Likewise, the function to cluster the
distance matrix and additional parameters can be specified in
<em>clust_function</em> and <em>clust_params</em>. By default, it uses
<code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">stats::hclust</a></code> with parameter
<code>method = "ward.D"</code>.</p></li>
<li><p>The order in which species are displayed can be defined by users
in parameter <em>cluster_species</em>. We strongly recommend passing a
vector of species order that matches the species tree, so that patterns
can be explored in a phylogenetic context. Importantly, if the character
vector is named, vector names will be used as species names in the plot.
This a nice way to replace species abbreviations with their full
names.</p></li>
</ol>
<p>Here, to briefly demonstrate how to play with the parameters we just
mentioned in the 3 remarks above, we will:</p>
<ul>
<li><p>Create a vector with the order in which we want species to be
displayed, with longer species names in vector names.</p></li>
<li><p>Create a metadata data frame containing the family of each
species.</p></li>
<li><p>Use the function <code>dsvdis()</code> from the <em><a href="https://CRAN.R-project.org/package=labdsv" class="external-link">labdsv</a></em> package
to calculate Ruzicka distances when clustering columns.</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1) Create a named vector of custom species order to plot</span></span>
<span><span class="va">species_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span></span>
<span>    <span class="co"># vector elements</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"vra"</span>, <span class="st">"van"</span>, <span class="st">"pvu"</span>, <span class="st">"gma"</span>, <span class="st">"cca"</span>, <span class="st">"tpr"</span>, <span class="st">"mtr"</span>, <span class="st">"adu"</span>, <span class="st">"lja"</span>,</span>
<span>        <span class="st">"Lang"</span>, <span class="st">"car"</span>, <span class="st">"pmu"</span>, <span class="st">"ppe"</span>, <span class="st">"pbr"</span>, <span class="st">"mdo"</span>, <span class="st">"roc"</span>, <span class="st">"fve"</span>,</span>
<span>        <span class="st">"Mnot"</span>, <span class="st">"Zjuj"</span>, <span class="st">"jcu"</span>, <span class="st">"mes"</span>, <span class="st">"rco"</span>, <span class="st">"lus"</span>, <span class="st">"ptr"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    <span class="co"># vector names</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"V. radiata"</span>, <span class="st">"V. angularis"</span>, <span class="st">"P. vulgaris"</span>, <span class="st">"G. max"</span>, <span class="st">"C. cajan"</span>,</span>
<span>        <span class="st">"T. pratense"</span>, <span class="st">"M. truncatula"</span>, <span class="st">"A. duranensis"</span>, <span class="st">"L. japonicus"</span>,</span>
<span>        <span class="st">"L. angustifolius"</span>, <span class="st">"C. arietinum"</span>, <span class="st">"P. mume"</span>, <span class="st">"P. persica"</span>,</span>
<span>        <span class="st">"P. bretschneideri"</span>, <span class="st">"M. domestica"</span>, <span class="st">"R. occidentalis"</span>, </span>
<span>        <span class="st">"F. vesca"</span>, <span class="st">"M. notabilis"</span>, <span class="st">"Z. jujuba"</span>,</span>
<span>        <span class="st">"J. curcas"</span>, <span class="st">"M. esculenta"</span>, <span class="st">"R. communis"</span>, </span>
<span>        <span class="st">"L. usitatissimum"</span>, <span class="st">"P. trichocarpa"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">species_order</span></span>
<span><span class="co">#&gt;        V. radiata      V. angularis       P. vulgaris            G. max </span></span>
<span><span class="co">#&gt;             "vra"             "van"             "pvu"             "gma" </span></span>
<span><span class="co">#&gt;          C. cajan       T. pratense     M. truncatula     A. duranensis </span></span>
<span><span class="co">#&gt;             "cca"             "tpr"             "mtr"             "adu" </span></span>
<span><span class="co">#&gt;      L. japonicus  L. angustifolius      C. arietinum           P. mume </span></span>
<span><span class="co">#&gt;             "lja"            "Lang"             "car"             "pmu" </span></span>
<span><span class="co">#&gt;        P. persica P. bretschneideri      M. domestica   R. occidentalis </span></span>
<span><span class="co">#&gt;             "ppe"             "pbr"             "mdo"             "roc" </span></span>
<span><span class="co">#&gt;          F. vesca      M. notabilis         Z. jujuba         J. curcas </span></span>
<span><span class="co">#&gt;             "fve"            "Mnot"            "Zjuj"             "jcu" </span></span>
<span><span class="co">#&gt;      M. esculenta       R. communis  L. usitatissimum    P. trichocarpa </span></span>
<span><span class="co">#&gt;             "mes"             "rco"             "lus"             "ptr"</span></span>
<span></span>
<span><span class="co"># 2) Create a metadata data frame containing the family of each species</span></span>
<span><span class="va">species_annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Species <span class="op">=</span> <span class="va">species_order</span>,</span>
<span>    Family <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Fabaceae"</span>, <span class="fl">11</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Rosaceae"</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>        <span class="st">"Moraceae"</span>, <span class="st">"Ramnaceae"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Euphorbiaceae"</span>, <span class="fl">3</span><span class="op">)</span>, </span>
<span>        <span class="st">"Linaceae"</span>, <span class="st">"Salicaceae"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">species_annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt;              Species   Family</span></span>
<span><span class="co">#&gt; V. radiata       vra Fabaceae</span></span>
<span><span class="co">#&gt; V. angularis     van Fabaceae</span></span>
<span><span class="co">#&gt; P. vulgaris      pvu Fabaceae</span></span>
<span><span class="co">#&gt; G. max           gma Fabaceae</span></span>
<span><span class="co">#&gt; C. cajan         cca Fabaceae</span></span>
<span><span class="co">#&gt; T. pratense      tpr Fabaceae</span></span>
<span></span>
<span></span>
<span><span class="co"># 3) Plot phylogenomic profiles, but using Ruzicka distances</span></span>
<span><span class="fu"><a href="../reference/plot_profiles.html">plot_profiles</a></span><span class="op">(</span></span>
<span>    <span class="va">profiles</span>, </span>
<span>    <span class="va">species_annotation</span>, </span>
<span>    cluster_species <span class="op">=</span> <span class="va">species_order</span>, </span>
<span>    dist_function <span class="op">=</span> <span class="fu">labdsv</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/labdsv/man/dsvdis.html" class="external-link">dsvdis</a></span>,</span>
<span>    dist_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>index <span class="op">=</span> <span class="st">"ruzicka"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="syntenet_files/figure-html/plot_profiles-1.png" width="700"></p>
<p>The heatmap is a nice way to observe patterns. For instance, you can
see some Rosaceae-specific clusters, Fabaceae-specific clusters, and
highly conserved ones as well.</p>
<p>If you want to explore in more details the group-specific clusters,
you can use the function <code><a href="../reference/find_GS_clusters.html">find_GS_clusters()</a></code>. For that, you
only need to input the profiles matrix and a data frame of species
annotation (i.e., species groups).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find group-specific clusters</span></span>
<span><span class="va">gs_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_GS_clusters.html">find_GS_clusters</a></span><span class="op">(</span><span class="va">profiles</span>, <span class="va">species_annotation</span><span class="op">)</span></span>
<span><span class="co">#&gt; Could not find annotation for species: </span></span>
<span><span class="co">#&gt; hlu</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gs_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt;        Group Percentage Cluster</span></span>
<span><span class="co">#&gt; 2   Fabaceae      90.91    1156</span></span>
<span><span class="co">#&gt; 21  Fabaceae      81.82    1170</span></span>
<span><span class="co">#&gt; 5  Ramnaceae     100.00    1279</span></span>
<span><span class="co">#&gt; 22  Fabaceae      90.91    1305</span></span>
<span><span class="co">#&gt; 23  Fabaceae      81.82    1309</span></span>
<span><span class="co">#&gt; 24  Fabaceae      90.91    1310</span></span>
<span></span>
<span><span class="co"># How many family-specific clusters are there?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gs_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 394</span></span></code></pre></div>
<p>As you can see, there are 394 family-specific clusters in the
network. Let’s plot a heatmap of group-specific clusters only.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter profiles matrix to only include group-specific clusters</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">profiles</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">gs_clusters</span><span class="op">$</span><span class="va">Cluster</span></span>
<span><span class="va">p_gs</span> <span class="op">&lt;-</span> <span class="va">profiles</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot heatmap</span></span>
<span><span class="fu"><a href="../reference/plot_profiles.html">plot_profiles</a></span><span class="op">(</span></span>
<span>    <span class="va">p_gs</span>, <span class="va">species_annotation</span>, </span>
<span>    cluster_species <span class="op">=</span> <span class="va">species_order</span>, </span>
<span>    cluster_columns <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="syntenet_files/figure-html/heatmap_filtered-1.png" width="700"></p>
<p>Pretty cool, huh? You can also visualize clusters as a network plot
with the function <code><a href="../reference/plot_network.html">plot_network()</a></code>. For example, let’s
visualize the group-specific clusters.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1) Visualize a network of first 5 GS-clusters</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="va">gs_clusters</span><span class="op">$</span><span class="va">Cluster</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plot_network.html">plot_network</a></span><span class="op">(</span><span class="va">network</span>, <span class="va">clusters</span>, cluster_id <span class="op">=</span> <span class="va">id</span><span class="op">)</span></span></code></pre></div>
<p><img src="syntenet_files/figure-html/plot_network-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># 2) Coloring nodes by family</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">network</span><span class="op">$</span><span class="va">node1</span>, <span class="va">network</span><span class="op">$</span><span class="va">node2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Gene <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    Species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">genes</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="va">head</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">gene_df</span>, <span class="va">species_annotation</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="st">"Family"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;              Gene   Family</span></span>
<span><span class="co">#&gt; 1 adu_Aradu.3W61T Fabaceae</span></span>
<span><span class="co">#&gt; 2 adu_Aradu.FQ8CE Fabaceae</span></span>
<span><span class="co">#&gt; 3 adu_Aradu.0W76I Fabaceae</span></span>
<span><span class="co">#&gt; 4 adu_Aradu.M4JP1 Fabaceae</span></span>
<span><span class="co">#&gt; 5 adu_Aradu.B26Z6 Fabaceae</span></span>
<span><span class="co">#&gt; 6 adu_Aradu.PLM7U Fabaceae</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_network.html">plot_network</a></span><span class="op">(</span><span class="va">network</span>, <span class="va">clusters</span>, cluster_id <span class="op">=</span> <span class="va">id</span>, color_by <span class="op">=</span> <span class="va">gene_df</span><span class="op">)</span></span></code></pre></div>
<p><img src="syntenet_files/figure-html/plot_network-2.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># 3) Interactive visualization (zoom out and in to explore it)</span></span>
<span><span class="fu"><a href="../reference/plot_network.html">plot_network</a></span><span class="op">(</span></span>
<span>    <span class="va">network</span>, <span class="va">clusters</span>, cluster_id <span class="op">=</span> <span class="va">id</span>, </span>
<span>    interactive <span class="op">=</span> <span class="cn">TRUE</span>, dim_interactive <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">300</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-5f00abdcc28102fc9c68" style="width:500px;height:300px;" class="forceNetwork html-widget"></div>
<script type="application/json" data-for="htmlwidget-5f00abdcc28102fc9c68">{"x":{"links":{"source":[3,0,2,5,0,2,6,4,3,1,6,8,2,0,7,5,7,0,13,5,4,9,3,6,2,8,12,5,13,14,10,1,11,9,3,4,15,9,15,4,3,19,1,10,18,3,15,19,9,20,4,10,21,18,6,12,8,2,16,0,14,13,7,5,17,8,24,2,6,16,12,19,20,3,15,4,22,23,21,1,18,10,6,26,8,16,24,2,12,0,25,14,13,7,5,17,4,20,3,9,15,22,19,10,23,21,1,18,28,23,18,1,28,10,32,21,8,2,6,12,29,16,24,26,3,31,20,9,22,19,4,15,27,30,5,25,13,14,33,23,32,21,10,1,18,28,26,6,8,24,16,34,29,2,12,22,3,9,31,27,35,20,19,4,15,34,24,16,38,26,2,29,6,8,12,5,14,36,25,13,30,18,33,10,28,21,23,37,1,32,43],"target":[4,5,6,7,7,8,8,9,9,10,12,12,12,13,13,13,14,14,14,14,15,15,15,16,16,16,16,17,17,17,18,18,18,19,19,19,19,20,20,20,20,20,21,21,21,22,22,22,22,22,22,23,23,23,24,24,24,24,24,25,25,25,25,25,25,26,26,26,26,26,26,27,27,27,27,27,27,28,28,28,28,28,29,29,29,29,29,29,29,30,30,30,30,30,30,30,31,31,31,31,31,31,31,32,32,32,32,32,32,33,33,33,33,33,33,33,34,34,34,34,34,34,34,34,35,35,35,35,35,35,35,35,35,36,36,36,36,36,37,37,37,37,37,37,37,37,38,38,38,38,38,38,38,38,38,39,39,39,39,39,39,39,39,39,39,40,40,40,40,40,40,40,40,40,40,41,41,41,41,41,41,42,42,42,42,42,42,42,42,42,44],"colour":["#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666","#666"]},"nodes":{"name":["adu_Aradu.AV6RW","adu_Aradu.E0LH8","adu_Aradu.JF3WA","adu_Aradu.R0IMB","car_03056.1","car_08943.1","car_19495.1","cca_02211","cca_04052","cca_05788","gma_03G097900","gma_07G124000","gma_07G238400","gma_10G041100","gma_13G128000","gma_14G033200","gma_17G035000","Lang_109346054","Lang_109352113","Lang_109355854","Lang_109359285","Lang_109362187","lja_2g3v3059790.1lja","lja_3g3v0642560.1lja","lja_4g3v2263980.1lja","mtr_1g069385","mtr_4g122270","mtr_5g090150","mtr_8g009900","pvu_003G116100","pvu_007G154300","pvu_008G199500","pvu_010G038400","tpr_gene17355","tpr_gene25402","van_01g01830.1","van_06g14500.1","van_08g06220.1","van_11g16300.1","vra_06g02600","vra_07g24530","vra_08g07810","vra_09g06430","Zjuj_107406488","Zjuj_107435072"],"group":["1309","1170","1305","1156","1156","1309","1305","1309","1305","1156","1170","1170","1305","1309","1309","1156","1305","1309","1170","1156","1156","1170","1156","1170","1305","1309","1305","1156","1170","1305","1309","1156","1170","1170","1305","1156","1309","1170","1305","1156","1305","1309","1170","1279","1279"],"nodesize":[" 6"," 8","10","10","10"," 9","10"," 6","10"," 9"," 9"," 1","10"," 9"," 9","10","10"," 5","10","10","10"," 9","10"," 8","10"," 9","10"," 8"," 9","10"," 9"," 9"," 9"," 9","10","10"," 6"," 9","10","10","10"," 6"," 9"," 1"," 1"]},"options":{"NodeID":"name","Group":"Group","colourScale":"d3.scaleOrdinal(d3.schemeCategory20);","fontSize":12,"fontFamily":"serif","clickTextSize":30,"linkDistance":50,"linkWidth":"function(d) { return Math.sqrt(d.value); }","charge":-10,"opacity":0.8,"zoom":true,"legend":false,"arrows":false,"nodesize":true,"radiusCalculation":" Math.sqrt(d.nodesize)+6","bounded":false,"opacityNoHover":0,"clickAction":null}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level2">
<h2 id="microsynteny-based-phylogeny-reconstruction">Microsynteny-based phylogeny reconstruction<a class="anchor" aria-label="anchor" href="#microsynteny-based-phylogeny-reconstruction"></a>
</h2>
<p>Finally, you can use the information on presence/absence of clusters
in each species to reconstruct a microsynteny-based phylogeny.</p>
<p>To do that, you first need to binarize the profiles matrix (0s and 1s
representing absence and presence, respectively) and transpose it. This
can be done with <code>binarize_and_tranpose()</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bt_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binarize_and_transpose.html">binarize_and_transpose</a></span><span class="op">(</span><span class="va">profiles</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Looking at the first 5 rows and 5 columns of the matrix</span></span>
<span><span class="va">bt_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      </span></span>
<span><span class="co">#&gt;       1 2 3 4 5</span></span>
<span><span class="co">#&gt;   adu 1 1 0 0 1</span></span>
<span><span class="co">#&gt;   car 1 1 1 1 1</span></span>
<span><span class="co">#&gt;   cca 1 1 1 1 1</span></span>
<span><span class="co">#&gt;   fve 1 1 1 1 1</span></span>
<span><span class="co">#&gt;   gma 1 1 1 1 1</span></span></code></pre></div>
<p>Now, you can use this transposed binary matrix as input to IQTREE2
<span class="citation">(Minh et al. 2020)</span> to infer a phylogeny.
To do so, you can use the function
<code><a href="../reference/infer_microsynteny_phylogeny.html">infer_microsynteny_phylogeny()</a></code>, which allows you to run
IQTREE2 from an R session <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Alternative:&lt;/strong&gt; if you want to use a
different program rather than IQTREE2, you can use the function
&lt;code&gt;&lt;a href="../reference/profiles2phylip.html"&gt;profiles2phylip()&lt;/a&gt;&lt;/code&gt; to write the transposed binary matrix to
a PHYLIP file and run your favorite program on the command line.
However, when inferring a phylogeny from phylogenomic profiles, you need
to make sure that the program you are using supports substitution models
for binary data. In IQTREE2, for instance, using binary, morphological
models requires passing parameters &lt;code&gt;-st MORPH&lt;/code&gt;.&lt;/p&gt;'><sup>4</sup></a>. You need to have IQTREE2 installed in your
machine and in your PATH to run this function. You can check if you have
IQTREE2 installed with <code><a href="../reference/iqtree_is_installed.html">iqtree_is_installed()</a></code>.</p>
<p>For the sake of demonstration, we will infer a phylogeny with
<code><a href="../reference/infer_microsynteny_phylogeny.html">infer_microsynteny_phylogeny()</a></code> using the profiles for BUSCO
genes for six legume species only. We will also remove non-variable
sites (i.e., clusters that are present in all species or absent in all
species). However, we’re only using this filtered data set for speed
issues. For real-life applications, the best thing to do is to
<strong>include phylogenomic profiles for all genes</strong> (not only
BUSCO genes).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Leave only 6 legume species and P. mume as an outgroup for testing purposes</span></span>
<span><span class="va">included</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gma"</span>, <span class="st">"pvu"</span>, <span class="st">"vra"</span>, <span class="st">"van"</span>, <span class="st">"cca"</span>, <span class="st">"pmu"</span><span class="op">)</span></span>
<span><span class="va">bt_mat</span> <span class="op">&lt;-</span> <span class="va">bt_mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">bt_mat</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">included</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Remove non-variable sites</span></span>
<span><span class="va">bt_mat</span> <span class="op">&lt;-</span> <span class="va">bt_mat</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">bt_mat</span><span class="op">)</span> <span class="op">!=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">included</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="../reference/iqtree_is_installed.html">iqtree_is_installed</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/infer_microsynteny_phylogeny.html">infer_microsynteny_phylogeny</a></span><span class="op">(</span><span class="va">bt_mat</span>, outgroup <span class="op">=</span> <span class="st">"pmu"</span>, </span>
<span>                                          threads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The output of <code><a href="../reference/infer_microsynteny_phylogeny.html">infer_microsynteny_phylogeny()</a></code> is a
character vector with paths to the output files from IQTREE2. Usually,
you are interested in the file ending in <em>.treefile</em>. This is the
species tree in Newick format, and it can be visualized with your
favorite tree viewer. We strongly recommend using the
<code><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree()</a></code> function from the Bioconductor package <em><a href="https://bioconductor.org/packages/3.15/treeio" class="external-link">treeio</a></em>
<span class="citation">(L.-G. Wang et al. 2020)</span> to read the tree,
and visualizing it with the <em><a href="https://bioconductor.org/packages/3.15/ggtree" class="external-link">ggtree</a></em>
Bioc package <span class="citation">(Yu et al. 2017)</span>. For
example, let’s visualize a microsynteny-based phylogeny for 123
angiosperm species, whose phylogenomic profiles were obtained from <span class="citation">Zhao et al. (2021)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">angiosperm_phylogeny</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting the tree</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.amazon.com/Integration-Manipulation-Visualization-Phylogenetic-Computational-ebook/dp/B0B5NLZR1Z/" class="external-link">ggtree</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/ggtree.html" class="external-link">ggtree</a></span><span class="op">(</span><span class="va">angiosperm_phylogeny</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggtree/man/geom_tiplab.html" class="external-link">geom_tiplab</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p><img src="syntenet_files/figure-html/vis_phylogeny-1.png" width="672"></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h2>
<div class="section level3">
<h3 class="unnumbered" id="how-do-i-execute-an-external-dependency-that-is-not-in-my-path">How do I execute an external dependency that is
not in my PATH?<a class="anchor" aria-label="anchor" href="#how-do-i-execute-an-external-dependency-that-is-not-in-my-path"></a>
</h3>
<p>If you have DIAMOND and/or IQTREE2 installed, but in a directory that
is not in your PATH, you can add this given directory to your PATH with
the function <code><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv()</a></code>.</p>
<p>For example, suppose your DIAMOND binary is in
<code>/home/username/bioinfo_tools</code>. To add this directory to your
PATH, you would run:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add example directory /home/username/bioinfo_tools to PATH</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span></span>
<span>    PATH <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"PATH"</span><span class="op">)</span>, <span class="st">"/home/username/bioinfo_tools"</span>, sep <span class="op">=</span> <span class="st">":"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that your R PATH is not the same as your system’s PATH. Thus,
even if you add the directory <code>/home/username/bioinfo_tools</code>
to your system’s path (e.g., by editing your ~/.bashrc file if you are
on Linux), you would still need to update your R PATH.</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was created under the following conditions:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggtree_3.4.4     syntenet_1.1.2   BiocStyle_2.24.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] nlme_3.1-160                labdsv_2.0-1               </span></span>
<span><span class="co">#&gt;   [3] bitops_1.0-7                matrixStats_0.62.0         </span></span>
<span><span class="co">#&gt;   [5] fs_1.5.2                    RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [7] rprojroot_2.0.3             GenomeInfoDb_1.32.4        </span></span>
<span><span class="co">#&gt;   [9] tools_4.2.1                 bslib_0.4.1                </span></span>
<span><span class="co">#&gt;  [11] utf8_1.2.2                  R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [13] lazyeval_0.2.2              mgcv_1.8-41                </span></span>
<span><span class="co">#&gt;  [15] BiocGenerics_0.42.0         colorspace_2.0-3           </span></span>
<span><span class="co">#&gt;  [17] withr_2.5.0                 tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt;  [19] compiler_4.2.1              textshaping_0.3.6          </span></span>
<span><span class="co">#&gt;  [21] cli_3.4.1                   Biobase_2.56.0             </span></span>
<span><span class="co">#&gt;  [23] intergraph_2.0-2            network_1.18.0             </span></span>
<span><span class="co">#&gt;  [25] desc_1.4.2                  DelayedArray_0.22.0        </span></span>
<span><span class="co">#&gt;  [27] labeling_0.4.2              rtracklayer_1.56.1         </span></span>
<span><span class="co">#&gt;  [29] bookdown_0.30               sass_0.4.2                 </span></span>
<span><span class="co">#&gt;  [31] scales_1.2.1                pkgdown_2.0.6              </span></span>
<span><span class="co">#&gt;  [33] yulab.utils_0.0.5           systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [35] stringr_1.4.1               digest_0.6.30              </span></span>
<span><span class="co">#&gt;  [37] Rsamtools_2.12.0            rmarkdown_2.18             </span></span>
<span><span class="co">#&gt;  [39] XVector_0.36.0              pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [41] htmltools_0.5.3             MatrixGenerics_1.8.1       </span></span>
<span><span class="co">#&gt;  [43] highr_0.9                   fastmap_1.1.0              </span></span>
<span><span class="co">#&gt;  [45] htmlwidgets_1.5.4           rlang_1.0.6                </span></span>
<span><span class="co">#&gt;  [47] gridGraphics_0.5-1          farver_2.1.1               </span></span>
<span><span class="co">#&gt;  [49] jquerylib_0.1.4             BiocIO_1.6.0               </span></span>
<span><span class="co">#&gt;  [51] generics_0.1.3              jsonlite_1.8.3             </span></span>
<span><span class="co">#&gt;  [53] statnet.common_4.7.0        BiocParallel_1.30.4        </span></span>
<span><span class="co">#&gt;  [55] dplyr_1.0.10                RCurl_1.98-1.9             </span></span>
<span><span class="co">#&gt;  [57] magrittr_2.0.3              ggplotify_0.1.0            </span></span>
<span><span class="co">#&gt;  [59] ggnetwork_0.5.10            GenomeInfoDbData_1.2.8     </span></span>
<span><span class="co">#&gt;  [61] patchwork_1.1.2             Matrix_1.5-3               </span></span>
<span><span class="co">#&gt;  [63] Rcpp_1.0.9                  munsell_0.5.0              </span></span>
<span><span class="co">#&gt;  [65] S4Vectors_0.34.0            fansi_1.0.3                </span></span>
<span><span class="co">#&gt;  [67] ape_5.6-2                   lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [69] stringi_1.7.8               yaml_2.3.6                 </span></span>
<span><span class="co">#&gt;  [71] MASS_7.3-58.1               SummarizedExperiment_1.26.1</span></span>
<span><span class="co">#&gt;  [73] zlibbioc_1.42.0             Rtsne_0.16                 </span></span>
<span><span class="co">#&gt;  [75] grid_4.2.1                  parallel_4.2.1             </span></span>
<span><span class="co">#&gt;  [77] crayon_1.5.2                lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [79] splines_4.2.1               Biostrings_2.64.1          </span></span>
<span><span class="co">#&gt;  [81] knitr_1.40                  pillar_1.8.1               </span></span>
<span><span class="co">#&gt;  [83] igraph_1.3.5                GenomicRanges_1.48.0       </span></span>
<span><span class="co">#&gt;  [85] rjson_0.2.21                codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [87] stats4_4.2.1                XML_3.99-0.12              </span></span>
<span><span class="co">#&gt;  [89] glue_1.6.2                  evaluate_0.18              </span></span>
<span><span class="co">#&gt;  [91] ggfun_0.0.8                 BiocManager_1.30.19        </span></span>
<span><span class="co">#&gt;  [93] treeio_1.20.2               vctrs_0.5.0                </span></span>
<span><span class="co">#&gt;  [95] networkD3_0.4               tidyr_1.2.1                </span></span>
<span><span class="co">#&gt;  [97] gtable_0.3.1                purrr_0.3.5                </span></span>
<span><span class="co">#&gt;  [99] cachem_1.0.6                ggplot2_3.4.0              </span></span>
<span><span class="co">#&gt; [101] xfun_0.34                   restfulr_0.0.15            </span></span>
<span><span class="co">#&gt; [103] tidytree_0.4.1              coda_0.19-4                </span></span>
<span><span class="co">#&gt; [105] ragg_1.2.4                  tibble_3.1.8               </span></span>
<span><span class="co">#&gt; [107] pheatmap_1.0.12             aplot_0.1.8                </span></span>
<span><span class="co">#&gt; [109] GenomicAlignments_1.32.1    memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [111] IRanges_2.30.1              cluster_2.1.4</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-altschul1997gapped" class="csl-entry">
Altschul, Stephen F, Thomas L Madden, Alejandro A Schäffer, Jinghui
Zhang, Zheng Zhang, Webb Miller, and David J Lipman. 1997. <span>“Gapped
BLAST and PSI-BLAST: A New Generation of Protein Database Search
Programs.”</span> <em>Nucleic Acids Research</em> 25 (17): 3389–3402.
</div>
<div id="ref-buchfink2021sensitive" class="csl-entry">
Buchfink, Benjamin, Klaus Reuter, and Hajk-Georg Drost. 2021.
<span>“Sensitive Protein Alignments at Tree-of-Life Scale Using
DIAMOND.”</span> <em>Nature Methods</em> 18 (4): 366–68.
</div>
<div id="ref-eddelbuettel2011rcpp" class="csl-entry">
Eddelbuettel, Dirk, and Romain François. 2011. <span>“Rcpp: Seamless r
and c++ Integration.”</span> <em>Journal of Statistical Software</em>
40: 1–18.
</div>
<div id="ref-lange2021cliquepercolation" class="csl-entry">
Lange, Jens. 2021. <span>“CliquePercolation: An r Package for Conducting
and Visualizing Results of the Clique Percolation Network Community
Detection Algorithm.”</span> <em>Journal of Open Source Software</em> 6
(62): 3210.
</div>
<div id="ref-minh2020iq" class="csl-entry">
Minh, Bui Quang, Heiko A Schmidt, Olga Chernomor, Dominik Schrempf,
Michael D Woodhams, Arndt Von Haeseler, and Robert Lanfear. 2020.
<span>“IQ-TREE 2: New Models and Efficient Methods for Phylogenetic
Inference in the Genomic Era.”</span> <em>Molecular Biology and
Evolution</em> 37 (5): 1530–34.
</div>
<div id="ref-vandepoele2013pico" class="csl-entry">
Vandepoele, Klaas, Michiel Van Bel, Guilhem Richard, Sofie Van
Landeghem, Bram Verhelst, Hervé Moreau, Yves Van de Peer, Nigel
Grimsley, and Gwenael Piganeau. 2013. <span>“Pico-PLAZA, a Genome
Database of Microbial Photosynthetic Eukaryotes.”</span>
<em>Environmental Microbiology</em> 15 (8): 2147–53.
</div>
<div id="ref-wang2020treeio" class="csl-entry">
Wang, Li-Gen, Tommy Tsan-Yuk Lam, Shuangbin Xu, Zehan Dai, Lang Zhou,
Tingze Feng, Pingfan Guo, et al. 2020. <span>“Treeio: An r Package for
Phylogenetic Tree Input and Output with Richly Annotated and Associated
Data.”</span> <em>Molecular Biology and Evolution</em> 37 (2): 599–603.
</div>
<div id="ref-wang2012mcscanx" class="csl-entry">
Wang, Yupeng, Haibao Tang, Jeremy D DeBarry, Xu Tan, Jingping Li, Xiyin
Wang, Tae-ho Lee, et al. 2012. <span>“MCScanX: A Toolkit for Detection
and Evolutionary Analysis of Gene Synteny and Collinearity.”</span>
<em>Nucleic Acids Research</em> 40 (7): e49–49.
</div>
<div id="ref-yu2017ggtree" class="csl-entry">
Yu, Guangchuang, David K Smith, Huachen Zhu, Yi Guan, and Tommy Tsan-Yuk
Lam. 2017. <span>“Ggtree: An r Package for Visualization and Annotation
of Phylogenetic Trees with Their Covariates and Other Associated
Data.”</span> <em>Methods in Ecology and Evolution</em> 8 (1): 28–36.
</div>
<div id="ref-zhao2017phylogenomic" class="csl-entry">
Zhao, Tao, Rens Holmer, Suzanne de Bruijn, Gerco C Angenent, Harrold A
van den Burg, and M Eric Schranz. 2017. <span>“Phylogenomic Synteny
Network Analysis of MADS-Box Transcription Factor Genes Reveals
Lineage-Specific Transpositions, Ancient Tandem Duplications, and Deep
Positional Conservation.”</span> <em>The Plant Cell</em> 29 (6):
1278–92.
</div>
<div id="ref-zhao2017network" class="csl-entry">
Zhao, Tao, and M Eric Schranz. 2017. <span>“Network Approaches for Plant
Phylogenomic Synteny Analysis.”</span> <em>Current Opinion in Plant
Biology</em> 36: 129–34.
</div>
<div id="ref-zhao2019network" class="csl-entry">
———. 2019. <span>“Network-Based Microsynteny Analysis Identifies Major
Differences and Genomic Outliers in Mammalian and Angiosperm
Genomes.”</span> <em>Proceedings of the National Academy of
Sciences</em> 116 (6): 2165–74.
</div>
<div id="ref-zhao2021whole" class="csl-entry">
Zhao, Tao, Arthur Zwaenepoel, Jia-Yu Xue, Shu-Min Kao, Zhen Li, M Eric
Schranz, and Yves Van de Peer. 2021. <span>“Whole-Genome
Microsynteny-Based Phylogeny of Angiosperms.”</span> <em>Nature
Communications</em> 12 (1): 1–14.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Fabrício Almeida-Silva, Tao Zhao, Kristian K Ullrich, Yves Van de Peer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
