# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @useDynLib syntenet, .registration = TRUE
#' @import Rcpp
#' @title rcpp_mcscanx_file
#' @name rcpp_mcscanx_file
#' @description MCSanX provides a clustering module for viewing the
#' relationship of colinear segments in multiple genomes (or heavily redundant
#' genomes). It takes the predicted pairwise segments from dynamic programming
#' (DAGchainer in particular) and then try to build consensus segments from a
#' set of related, overlapping segments.
#' @return list
#' @param blast_file blast input
#' @param gff_file gff input
#' @param prefix output prefix (default: out)
#' @param outdir output directory (default: "")
#' @param match_score match score (default: 50)
#' @param gap_penalty gap penalty (default: -1)
#' @param match_size match_size (default: 5)
#' @param e_value e_value (default: 1e-5)
#' @param max_gaps max gaps (default: 25)
#' @param overlap_window overlap window (default: 5)
#' @param is_pairwise specify if only pairwise blocks should be reported
#' (default: FALSE)
#' @param is_synteny specify patterns of collinear blocks.
#' 0: intra- and inter-species (default); 1: intra-species; 2: inter-species
#' @references Wang et al. (2012) MCScanX: a toolkit for detection and
#' evolutionary analysis of gene synteny and collinearity.
#' \emph{Nucleic acids research}. \bold{40.7}, e49-e49.
#' @references Haas et al. (2004) DAGchainer: a tool for mining segmental
#' genome duplications and synteny. \emph{Bioinformatics}. \bold{20.18}
#' 3643-3646.
#' @export rcpp_mcscanx_file
#' @author Kristian K Ullrich
rcpp_mcscanx_file <- function(blast_file, gff_file, prefix = "out", outdir = "", match_score = 50L, gap_penalty = -1L, match_size = 5L, e_value = 1e-5, max_gaps = 25L, overlap_window = 5L, is_pairwise = FALSE, in_synteny = 0L) {
    .Call(`_syntenet_rcpp_mcscanx_file`, blast_file, gff_file, prefix, outdir, match_score, gap_penalty, match_size, e_value, max_gaps, overlap_window, is_pairwise, in_synteny)
}

